### ISSUE TITLE
High gravity/jump launch sends player through platforms

### ISSUE THREAD

**Issue by stefankendall** [britzl/platypus#13]
When setting a high launch speed or gravity, a game object is able to move through a platform going up (jumping), or through a platform going down (falling).

https://github.com/stefankendall/platform

https://github.com/stefankendall/platform/blob/master/high_gravity.mp4?raw=true

**Comment by britzl** [comment]
Thank you for reporting a problem. I took a look at your project and was able to reproduce the problem. This is what I did to debug:

* Enabled physics debug in game.project to see collision shapes:

![Screenshot 2019-08-02 at 11 26 05](https://user-images.githubusercontent.com/1300688/62360166-84996a00-b518-11e9-9098-c91a7c01f549.png)

* Next I enabled physics debug in Platypus by passing in debug = true in the call to platypus.create()

```
    this.platypus = platypus.create({
        collisions = {
            separation = platypus.SEPARATION_SHAPES,
            groups = {
                [hash("ground")] = platypus.DIR_ALL,
            },
            left = 7, right = 7, top = 10, bottom = 10,
        },
        max_velocity = 2000,
        gravity = kGravity,
        allow_double_jump = false,
        allow_wall_jump = false,
        debug = true,
    })
```

Results in:

![Screenshot 2019-08-02 at 11 26 31](https://user-images.githubusercontent.com/1300688/62360272-b6aacc00-b518-11e9-86c1-f89ee4e030c8.png)

* From the config and with physics debug enabled we see that the ray casts done to detect ground collisions etc are really tiny:

![Screenshot 2019-08-02 at 11 30 00](https://user-images.githubusercontent.com/1300688/62360413-038ea280-b519-11e9-9119-bb2f36f9debe.png)

The size corresponds to left, right, top and bottom values passed into the config. The values should match the size of the collision shape. This is especially important since you have set collision resolution to happen using the collision shape (you could also resolve collisions based just on the result of the ray casts).

* If we increase the left, right, top and bottom values to match the size of the shape we get this:

```
            left = 54, right = 54, top = 75, bottom = 75,
```

<img width="223" alt="Screenshot 2019-08-02 at 11 33 54" src="https://user-images.githubusercontent.com/1300688/62360590-6f710b00-b519-11e9-884d-693f29942993.png">

This is better but obviously not right. It highlights the fact that I have assumed that the collision shape and sprites are centred around 0x0 while in your case the sprite is offset. If we change the left, right, top, bottom values to match we get this:

```
            left = 54, right = 54, top = 108, bottom = 1,
```

But it doesn't work. The player falls through the ground. This is because the ray casts pointing down, down left and down right aren't angled properly and the one pointing straight down is almost 0 in length. Increasing bottom gives a better result:

```
            left = 54, right = 54, top = 108, bottom = 10,
```

<img width="168" alt="Screenshot 2019-08-02 at 11 39 24" src="https://user-images.githubusercontent.com/1300688/62360921-371dfc80-b51a-11e9-9249-22592a5f6b14.png">

But the player is hovering above the ground. Not good. There are two alternatives now:

1. I fix this by adding a new config value called offset (vector3) which is used when doing ray casts. This value would offset the raycast start position so that it is possible to have sprites and shapes offset from 0x0 on the game object.
2. Move your sprite and shape to solve the problem:

![Screenshot 2019-08-02 at 11 43 50](https://user-images.githubusercontent.com/1300688/62361243-e2c74c80-b51a-11e9-8cf8-52ec89a8a403.png)

![Screenshot 2019-08-02 at 11 44 07](https://user-images.githubusercontent.com/1300688/62361244-e35fe300-b51a-11e9-8f62-1ef98c3b919b.png)

**Comment by britzl** [comment]
Released 3.2.1 with this new config option.

https://github.com/britzl/platypus/releases/tag/3.2.1

See grottoescape game in project for example of use.
